### ISSUE TITLE
Typewriter example

### ISSUE THREAD

**Issue by amyjko** [wordplaydev/wordplay#432]
## What's the problem?

There aren't very many writing related examples in the galleries at the moment; it seems like a missing genre that might resonate with youth who want to create more writing-oriented applications. One fun idea might be a typewriter like application, for writing messages in typewriter style.

## What's the design idea?

The essence of the idea would be to have a blank canvas and a cursor and that keystrokes would be typed onto the screen, with enter keys creating new lines and backspace deleting characters. There could be fun animations to simulate the head hitting paper.

## Design specification

_Reporters can leave this blank. Designers should write a detailed, precise description of what is to be built. If images are necessary to convey this precisely, make sure to include image descriptions, so everyone can see what is in the image. Do not link to external documents; everything should be embedded here. Once the design is approved, we will remove the `needs design` tag and replace it with the `buildable` tag, signalling that it can be built._

**Comment by velazjoc** [comment]
Hi @amyjko this seems interesting to start, is there any way I could get assigned to this?

**Comment by amyjko** [comment]
Assigned!

**Comment by ericazqyy** [comment]
Hi @amyjko, I am also pretty interested in this issue. Can I elaborate with @velazjoc?

**Comment by ericazqyy** [comment]
Hi @amyjko, I am also pretty interested in this issue. Can I collaborate with @velazjoc?

**Comment by amyjko** [comment]
You should ask @velazjoc! You're welcome to collaborate together if you're both agreeable to it.

**Comment by joonpiter** [comment]
Hey! @velazjoc and @ericazqyy I would like to start working on this issue as well with yall

**Comment by ericazqyy** [comment]
Hi Isabel,

Welcome! We were wondering if you are in section today, so we can work on
it together!

Best,
Erica Zhang

**Comment by velazjoc** [comment]
> Hey! @velazjoc and @ericazqyy I would like to start working on this issue as well with yall

@joonpiter Hi! We're in section today towards the front, but if you aren't here today just give us your discord username so we can add you to the groupchat!

**Comment by joonpiter** [comment]
Hi, for some reason I wasnt pinged in your comments, ill just work on reviewing your design for now!

**Comment by joonpiter** [comment]
- The proposal mentions that letters will "fade out" using an opacity animation. How long will this animation take, and will it affect the readability of longer texts if they start fading too quickly?
- You state that new lines will begin once a line limit is reached based on font size. Is there a specific algorithm or rule for determining this limit? What happens if the font size changes mid-line due to the typewriter's "naughty" behavior?
-  Jamming will occur randomly, requiring 2 to 10 presses for a character to appear. How frequently should this jamming behavior occur to balance frustration and fun?
- The typewriter will randomly insert characters and change font sizes/colors. How frequently will these random insertions occur? Is there a cap on how many random characters can be inserted in a single sentence or document?
-  Finally, the proposal outlines random color and size changes. Will a consistent set of colors and sizes exist, or will these be entirely random? How will accessibility be maintained if the colors contrast poorly with the background?

**Comment by iamvivian03** [comment]
> * The proposal mentions that letters will "fade out" using an opacity animation. How long will this animation take, and will it affect the readability of longer texts if they start fading too quickly?
> * You state that new lines will begin once a line limit is reached based on font size. Is there a specific algorithm or rule for determining this limit? What happens if the font size changes mid-line due to the typewriter's "naughty" behavior?
> * Jamming will occur randomly, requiring 2 to 10 presses for a character to appear. How frequently should this jamming behavior occur to balance frustration and fun?
> * The typewriter will randomly insert characters and change font sizes/colors. How frequently will these random insertions occur? Is there a cap on how many random characters can be inserted in a single sentence or document?
> * Finally, the proposal outlines random color and size changes. Will a consistent set of colors and sizes exist, or will these be entirely random? How will accessibility be maintained if the colors contrast poorly with the background?

Hi @joonpiter! Here's our brainstorming Ideas for these questions:
> * The proposal mentions that letters will "fade out" using an opacity animation. How long will this animation take, and will it affect the readability of longer texts if they start fading too quickly?

- For an actual typewriter, pressing harder results in darker text. However, it’s hard to track how hard the user presses. Thus, in this design, we propose incorporating a random fadeout effect on the text. This effect will not be animated but will occur immediately after the user types.

> * You state that new lines will begin once a line limit is reached based on font size. Is there a specific algorithm or rule for determining this limit? What happens if the font size changes mid-line due to the typewriter's "naughty" behavior?

- We decided on this typewriter rule that users are not allowed to change the font size when they start working on one paper, which means the font size for one single paper is fixed. 
- For the code logic, we could first set a fixed line length. Then, we can use the algorithm to keep track of the total number of letters being pressed on one single line and use that total number of letters to time the width of the current font size to determine whether the line limit has been reached. 

> * Jamming will occur randomly, requiring 2 to 10 presses for a character to appear. How frequently should this jamming behavior occur to balance frustration and fun?

- Normally, typewriters “jam” when too many keys are pressed at the same time. However, for simplicity of code, we decided that 2 to 6 presses will be required instead, with the amount of presses required to print the character being chosen randomly. The jamming behavior should only occur after 15 characters have been typed, in order for the user to get used to typing. 
- After the first 15 characters have been typed, any character from then on may get jammed. However, the amount of times this occurs is selected randomly, anywhere from 0 to 15 times on any random character. Since it is random, there is a possibility that a character will have the jamming behavior right after another just did.

> * The typewriter will randomly insert characters and change font sizes/colors. How frequently will these random insertions occur? Is there a cap on how many random characters can be inserted in a single sentence or document?

- We have decided the frequency to be 1~2/10. So, the total number of random characters that can be inserted is based on the lines users have typed. 

> * Finally, the proposal outlines random color and size changes. Will a consistent set of colors and sizes exist, or will these be entirely random? How will accessibility be maintained if the colors contrast poorly with the background?
- After consideration, we have decided to stick to random color changes, and the size will remain consistent throughout the page. 
- There will be a consistent set of colors that follows the WCAG to be picked from randomly.

**Comment by joonpiter** [comment]
@iamvivian03, thanks for the response. Would you edit your proposal to reflect your current decisions, and then we can revise from there?

**Comment by ericazqyy** [comment]
> @iamvivian03, thanks for the response. Would you edit your proposal to reflect your current decisions, and then we can revise from there?

Yes, I have updated the proposal with our current decisions along with some visualization examples!

**Comment by velazjoc** [comment]
Status: Open issue
Work Completed: We began the basic design requirements and initial proposal of a writing-oriented typewriter gallery example. Our research consisted of understanding what vintage and modern typewriter's looked and behaved like in order to build up the basic typewriter visualization. Our visualization examples showcase what it could look like. Furthermore, we outlined the behavior of a fun and silly "Wordplay" typewriter that misbehaves by refusing to act like a normal typewriter. Our proposal is only the start of an idea and can be modified to become more feasible.
Materials: Research and figma designs have been included as text and photos in the main body issue, as well as the specific design requirements to create the typewriter.
Contributers: @ericazqyy @iamvivian03 @yishiz2 @velazjoc

**Comment by amyjko** [comment]
@velazjoc @ericazqyy @iamvivian03 @yishiz2, thank you for the update. Are you continuing your work on this? If not, please unassign yourselves, so everyone knows.

**Comment by iamvivian03** [comment]
Hi @amyjko @joonpiter @thesmolcoder,
I just updated the design proposal, with the main changes in the Design Specification and Visualization Examples sections.

**Comment by amyjko** [comment]
Nice update! Overall, it seems like a fun experience. A few thoughts:

* Wordplay doesn't support and isn't designed for creating graphics. But the design involves many graphics. (Paper, button chrome, etc). If you haven't yet, learn the platform and it's limitations before finalizing your design, so you're sure it's possible to create.
* Why have an on screen keyboard? This limits the characters that can be typed on the keyboard and imposes accessibility challenges on anyone that cannot use a pointing device, slowing text entry. Perhaps your rationale was to closely mimic a keyboard for English? Since the goal of Wordplay is to support all of the world's languages (and making an on screen keyboard for all languages would be wayyyyy to much work), and another goal is accessibility, these design choices seem in conflict with the goals of the platform. Why not just rely on the operating system for text entry, using whatever input device it supports?

**Comment by iamvivian03** [comment]
Hello @amyjko, I've updated the revised design, fixing the issue from before. The design looks complete unless there are any functional issues.

**Comment by amyjko** [comment]
@iamvivian03 That's great! Will you also be implementing this example?

**Comment by iamvivian03** [comment]
> @iamvivian03 That's great! Will you also be implementing this example?

Status: Open issue(design completed, need implementing)
I won’t be continuing with this next quarter. I'll unassign myself to make it available for others to assign.
